include/unifex/materialize.hpp

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/async_trace.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <type_traits>
+
+namespace unifex
+{
+  namespace detail
+  {
+    template <typename Receiver>
+    class materialize_receiver {
+    public:
+      template <typename Receiver2>
+      explicit materialize_receiver(Receiver2&& receiver) noexcept(
+          std::is_nothrow_constructible_v<Receiver, Receiver2>)
+        : receiver_(static_cast<Receiver2&&>(receiver)) {}
+
+      template <
+          typename... Values,
+          std::enable_if_t<
+              std::is_invocable_v<
+                  decltype(unifex::set_value),
+                  Receiver,
+                  decltype(unifex::set_value),
+                  Values...>,
+              int> = 0>
+      void
+      set_value(Values&&... values) && noexcept(std::is_nothrow_invocable_v<
+                                                decltype(unifex::set_value),
+                                                Receiver,
+                                                decltype(unifex::set_value),
+                                                Values...>) {
+        unifex::set_value(
+            static_cast<Receiver&&>(receiver_),
+            unifex::set_value,
+            static_cast<Values&&>(values)...);
+      }
+
+      template <
+          typename Error,
+          std::enable_if_t<
+              std::is_invocable_v<
+                  decltype(unifex::set_value),
+                  Receiver,
+                  decltype(unifex::set_error),
+                  Error>,
+              int> = 0>
+      void set_error(Error&& error) && noexcept {
+        if constexpr (std::is_nothrow_invocable_v<
+                          decltype(unifex::set_value),
+                          Receiver,
+                          decltype(unifex::set_error),
```

Comment:
Why `decltype` here instead of `tag_t`?

Reply:
`tag_t` returns the decayed type.
Whereas when we invoke the CPO we are invoking it with the CPO objects themselves which will be const-lvalue-ref.

Reply:
Maybe this should be `decltype((unifex::set_value))` instead to actually have it decay to the type of the expression rather than the type of the declaration?